src/datasets/arrow_dataset.py

```diff
@@ -159,6 +159,131 @@ def version(self):
+                    "sure the column should be convertable to tf.Tensor, please "
+                    "file an issue at github.com/huggingface/datasets and tag "
+                    "@rocketknight1!"
+                )
+            shape = [dim if dim != -1 else None for dim in shape]
+
+            signatures[column] = tf.TensorSpec(shape=shape, dtype=dtype)
+        return signatures
+
+    def to_tf_dataset(self, columns, batch_size, shuffle, collate_fn=None, label_cols=None, pad_to=0, prefetch=True):
+        import tensorflow as tf
+
+        if label_cols is None:
+            label_cols = []
+        elif isinstance(label_cols, str):
+            label_cols = [label_cols]
+        elif len(set(label_cols)) < len(label_cols):
+            raise ValueError("List of label_cols contains duplicates!")
+        if not columns:
+            raise ValueError("Need to specify at least one column!")
+        elif isinstance(columns, str):
+            columns = [columns]
+        elif len(set(columns)) < len(columns):
+            raise ValueError("List of columns contains duplicates!")
+        if pad_to > 0 and collate_fn is not None:
+            raise ValueError(
+                "pad_to cannot be used with a custom collate_fn - you should modify your collate_fn instead!"
+            )
+        if label_cols is not None:
+            cols_to_retain = list(set(columns + label_cols))
+        else:
+            cols_to_retain = columns
+        for col in cols_to_retain:
+            if col not in self.features:
+                raise ValueError(f"Couldn't find column {col} in dataset!")
+        dataset = self.remove_columns([col for col in self.features if col not in cols_to_retain])
+        gen_signature = self._get_output_signature(dataset, batch_size)
+        num_batches = len(dataset) // batch_size  # Because we drop the remainder
+
+        def tf_generator():
+            # Note that the 'tensorflow' return format uses ragged tensors, which are VERY unperformant
+            # right now (TF 2.5). This may or may not change in the future, but for now we stick to 'numpy'.
+            if shuffle:
+                epoch_dataset = dataset.shuffle(load_from_cache_file=False)
+            else:
+                epoch_dataset = dataset
+            if collate_fn is None:
+                epoch_dataset.set_format("numpy")  # Automatic padding
+            else:
+                epoch_dataset.set_format("python")  # List of possibly variable lists
+            for i in range(0, len(epoch_dataset) - batch_size + 1, batch_size):
+                batch = epoch_dataset[i : i + batch_size]
+                if collate_fn is not None:
+                    batch = collate_fn(batch)
+                    batch = {key: np.array(val) for key, val in batch.items()}
+                yield batch
+
+        tf_dataset = tf.data.Dataset.from_generator(tf_generator, output_signature=gen_signature)
+
+        if pad_to > 0:
```

Comment:
If we include padding here, it would be nice to have the option to pad to max length of the batch, not a fixed number. And it also should deal with tensors that are too long and truncate them? Also why always pad on the last dimension and not another one?

This is all way too specific in my opinion, so I stand with just leaving a generic collate function, that can default to something that convert to tensors (and only work for things that are already of the same size), leaving any other use case up to the user.

Reply:
One thing to note here is that when we fetch numpy/torch tensors from our datasets, they automatically pad with zeroes on the right up to the maximum size in the batch. So maybe the easiest thing to do is to just rely on that behaviour when the user doesn't supply a collate_fn, and leave out any pad arguments?

Reply:
Oh? I have never seen that automatic padding in PyTorch, I get errors but maybe it's a recent thing.

Reply:
Maybe it's just Numpy, but I swear I remember testing with torch outputs too!

Reply:
Either way, padding is now removed outside of the user-specified `collate_fn`